### PR TITLE
Reset query and viz

### LINKED ISSUE #2916: Data / Button to reset query & vizz
image

A user must be able to 

- [ ] reset the query
- [ ] reset the vizz

### PR DESCRIPTION
**Pull request by jorgeatgu**
Closes #2916 #2970

## :v: What does this PR do?

Include button for reset query from editor.
Include button for reset Visualization.

## :mag: How should this be manually tested?
Staging

Steps to reset query:

1. Edit the query on editor.
2. Click on the reset query button.
3. The default query will appear in the editor

Steps to reset saved query:

1. Load a saved query.
2. Edit the saved query on editor.
3. Click on the reset query button.
4. The saved query will appear in the editor.

Steps to reset Visualization:

1. Click on Visualize/Visualizar
2. Select a new type of Visuzaliton.
3. Saved Visualization and reset visualization buttons will appear, the Visualize button will disappear.
4. Click on the reset visualization button.
5. Perspective visualization changes to Grid, and Perspective menus shouldn't be seen.

### CODE REVIEW

**Comment by furilo** [comment]
We should always show the button but show it not active if it's not active (for example, when you initially enter the data tab, the button should appear with opacity 50%)

**Comment by furilo** [comment]
Things to do:

- [x] Home button in query should (1) reset to the default query or (2) take you to /editor
- [x] In a Saved Query: recent and queries buttons should be compact (review design)
- [x] In a Saved Query, if query not modified: show button "Edit" (instead of Save -- as it was originally in the design https://www.figma.com/proto/IFqcY4iCKC9ukJkgfSRKEA/1910-Gobierto-Datos-Abiertos-Bocetos?node-id=100%3A0&viewport=588%2C-8806%2C1&scaling=min-zoom) (we talked about removing this button, but in order to not introduce more changes, let keep it this way)
- [x] In a Saved Query, if query is modified show a "revertir" link. Design https://www.figma.com/proto/IFqcY4iCKC9ukJkgfSRKEA/1910-Gobierto-Datos-Abiertos-Bocetos?node-id=100%3A506&viewport=588%2C-8806%2C1&scaling=min-zoom (also, Save button should be after the "Consulta modificada")

After doing this, and a thorough review, check that these no longer happens: 

- [x] If I load a Saved Query (ie. https://getafe.gobify.net/datos/censo-secciones/q/123) I'm shown only the default "Save" button, but I don't see the query name. 
- [x] If I'm in a Saved Query, edit the query, and click in Home button, I keep seeing the "consulta modificada" warn, and if my query has been reseted it is no longer modified thus I should not see that.

File: app/javascript/gobierto_data/lib/router.js [hunk 1]
```diff
@@ -27,7 +27,7 @@ export const router = new VueRouter({
       }, {
         path: ":id/q/:queryId?",
         component: Dataset,
-        name: 'Dataset',
+        name: 'Query',
```
Comment (Crashillo) [inline app/javascript/gobierto_data/lib/router.js:30]: Why so? If you do changes like that, rename ALL stuff accordingly and provide an explanation

File: app/javascript/gobierto_data/webapp/components/Dataset.vue [hunk 2]
```diff
@@ -149,20 +153,17 @@ export default {
 
         this.parseUrl({ queryId, sql });
       }
+
+      if (to.path !== from.path) {
+        this.setDefaultQuery()
+      }
+
+      if (to.name === 'Query') {
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/Dataset.vue]: Let's try to avoid *named* variables in the code

File: app/javascript/gobierto_data/webapp/components/Dataset.vue [hunk 3]
```diff
@@ -446,6 +459,12 @@ export default {
       // TODO: indicar algo con el status OK
       console.log("postVisualization", status);
     },
+    getColumnsQuery(csv) {
+      const lines = csv.split("\n");
+
+      const columns = lines[0].split(",");
+      this.arrayColumnsQuery = columns
+    }
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/Dataset.vue]: Data validation:

```suggestion
    getColumnsQuery(csv = '') { // it could come empty
      const [ columns = '' ] = csv.split("\n"); // access to the first array position, if not exists, set to empty string
      this.arrayColumnsQuery = columns.split(",") // split works for empty strings
    }
```

File: app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue [hunk 4]
```diff
@@ -100,19 +100,29 @@ export default {
...
       labelCancel: I18n.t('gobierto_data.projects.cancel') || "",
       labelEdit: I18n.t("gobierto_data.projects.edit") || ""
     }
   },
+  mounted() {
+    this.$root.$on('enableSavedButton', this.enableSavedButton)
+  },
+  beforeDestroy() {
+    this.$root.$off('enableSavedButton')
+  },
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue]: Stop doing this. Forbidden. A component never must listen events. You're binding the component to the content.
Use props instead. A component is updated through properties. The container (Dataset) listen the events and populate its children with the new changes.

File: app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue [hunk 5]
```diff
@@ -141,6 +152,9 @@ export default {
       const { checked } = event.target
       this.isPrivate = checked
     },
+    enableSavedButton() {
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue]: Very likely it becomes deprecated once you remove the events

File: app/javascript/gobierto_data/webapp/components/commons/Visualizations.vue [hunk 6]
```diff
@@ -44,6 +65,37 @@ export default {
...
+      const shadowRootPerspective = document.querySelector('perspective-viewer').shadowRoot
+      const configButtonPerspective = shadowRootPerspective.getElementById('config_button')
+      configButtonPerspective.style.display = "none"
+      const selectVizPerspective = shadowRootPerspective.getElementById('vis_selector')
+      const showDialog = this.showSavingDialog
+      const sendSelectedValue = this.selectedValue
+      let selectedValue
+
+      selectVizPerspective.addEventListener('change', function() {
+        selectedValue = selectVizPerspective.options[selectVizPerspective.selectedIndex].value;
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/Visualizations.vue]: Are you enough sure this object never could come `null`?

File: app/javascript/gobierto_data/webapp/components/commons/Visualizations.vue [hunk 7]
```diff
@@ -44,6 +65,37 @@ export default {
...
+        showDialog()
+        sendSelectedValue(selectedValue)
+      })
+    },
+    showSavingDialog() {
+      this.$emit("showSaving")
+    },
+    selectedValue(chart) {
+      this.$emit("selectedChart", chart)
+    },
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/Visualizations.vue]: 1. `showDialog`: you are creating a variable, to store a function to emit later an event. Nonsense, quite hard to follow/understand.
2. `selectedValue`: you create a variable uninitializated, another to store a function with the same name as the first variable; then set the value in the event to emit it in the function. WTF.

Just emit both events inside the listener. Simple as that.

```suggestion
      selectVizPerspective.addEventListener('change', () => {
        selectedValue = selectVizPerspective.options[selectVizPerspective.selectedIndex].value;
        
        this.$emit("showSaving")
        this.$emit("selectedChart", selectedValue)
      })
    },
```

If you need to keep the `this` context, just use an arrow function.

When the `change` event is triggered? Why do we need two events to handle this?

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorCode.vue [hunk 8]
```diff
@@ -107,6 +111,7 @@ export default {
     // metaKey + keyUp doesn't work with MacOS
     // https://stackoverflow.com/questions/11818637/why-does-javascript-drop-keyup-events-when-the-metakey-is-pressed-on-mac-browser
     this.editor.on("keydown", this.onKeyDown);
+    this.$root.$on('resetQuery', this.resetEditor)
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorCode.vue]: Once again, listeners everywhere. Why? why? why? Use `watched` properties

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorResults.vue [hunk 9]
```diff
@@ -59,6 +115,40 @@ export default {
...
+
+      this.$refs.viewer.enableDisabledPerspective(hidePerspective);
+      this.$refs.viewer.setColumns();
+    },
+    showChart() {
+      const showPerspective = "flex"
+
+      this.showVisualization = true
+      this.$refs.viewer.enableDisabledPerspective(showPerspective);
+    },
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorResults.vue:139]: What are this Visualizations functions doing enclosed here? You're messing with the logic. SQLEditorResults does know nothing about what Visualizations does. The parent just "inform" the children about showing/hiding, nothing else.
These functions does not belong here.

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorResults.vue [hunk 10]
```diff
@@ -59,6 +115,40 @@ export default {
...
+    showSavingDialog() {
+      this.perspectiveChanged = true
+      this.showVisualize = false
+      this.showResetViz = true
+      this.isVisualizationModified = true
+    },
+    changeChart(chart) {
+      this.typeChart = chart
+    },
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorResults.vue]: You may simplify this function straight out the DOM.

File: app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue [hunk 11]
```diff
@@ -111,6 +95,27 @@ export default {
       }
     }
   },
+  watch: {
+    currentDatasetSlug: function(newSlug) {
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue]: Use ES6 syntax

File: app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue [hunk 12]
```diff
@@ -111,6 +95,27 @@ export default {
...
+      this.showLess = true;
+      if (this.sortedItems.length && newSlug !== '') {
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue]: If you only wanna check if `newSlug` has value, leave this:
```suggestion
      if (this.sortedItems.length && newSlug) {
```

File: app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue [hunk 13]
```diff
@@ -111,6 +95,27 @@ export default {
...
+        this.showToggle = Object.keys(this.activeDatasetColumns).length > this.showMaxKeys
+      }
+    }
+  },
+  created() {
+    this.labelSets = I18n.t("gobierto_data.projects.sets")
+    this.labelQueries = I18n.t("gobierto_data.projects.queries")
+    this.labelCategories = I18n.t("gobierto_data.projects.categories")
+    this.labelshowAll = I18n.t("gobierto_data.projects.showAll")
+    this.labelshowLess = I18n.t("gobierto_data.projects.showLess")
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sidebar/SidebarDatasets.vue]: Move these vars to the `data()`

File: config/locales/gobierto_data/views/ca.yml [hunk 14]
```diff
@@ -43,16 +43,20 @@ ca:
...
+      modifiedVisualization: Visualització modificada
       news: Notícies
       private: Privat
       queries: Consultes
       queryExecuted: Consulta executada en
       queryName: Nom consulta
       recents: Recents
       records: Registres
+      resetQuery: Restableix la consulta
+      resetViz: Restableix la Visualització
```
Comment (Crashillo) [inline config/locales/gobierto_data/views/ca.yml]: Keep consistency between languages:
```suggestion
      resetQuery: Restablir consulta
      resetViz: Restablir Visualització
```

**Review by Crashillo — CHANGES_REQUESTED** [review]

File: app/javascript/gobierto_data/lib/router.js (hunk 1 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_data/lib/router.js:30]: Because we only use this route to render a saved query, also I need one way to see when the user loads an entire Dataset or only load a saved query belongs to Dataset.

File: app/javascript/gobierto_data/webapp/components/commons/Visualizations.vue (hunk 6 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_data/webapp/components/commons/Visualizations.vue]: Yes, these options are part of Perspective

File: app/javascript/gobierto_data/webapp/components/Dataset.vue [hunk 15]
```diff
@@ -221,7 +254,9 @@ export default {
       // update the editor text content by default
       this.currentQuery = `SELECT * FROM ${this.tableName} LIMIT 50`;
     }
+
     this.runCurrentQuery();
+    this.setDefaultQuery();
 
   },
   mounted() {
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/Dataset.vue:255]: Why not in `activated`?

File: app/javascript/gobierto_data/webapp/components/Dataset.vue [hunk 16]
```diff
@@ -366,6 +425,9 @@ export default {
         const { queryId } = this.$route.params;
         // factory method
         ({ status } = await this.putQuery(queryId, { data }));
+
+        //Update revert query
+        this.queryRevert = this.currentQuery
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/Dataset.vue:445]: Delete this

File: app/javascript/gobierto_data/webapp/components/Dataset.vue [hunk 17]
```diff
@@ -375,11 +437,31 @@ export default {
...
+
+      const {
+        id: queryId
+      } = lastQuery;
+
+      this.$router.push(
+        `/datos/${this.$route.params.id}/q/${queryId}`
+      // eslint-disable-next-line no-unused-vars
+      ).catch(err => {})
+    },
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/Dataset.vue]: Remove this

File: app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue [hunk 18]
```diff
@@ -37,43 +36,48 @@
...
-        color="var(--color-base)"
-        background="#fff"
-        class="btn-sql-editor"
-        @click.native="onClickSaveHandler"
-      />
+
+    <template v-if="isQuerySaved">
+      <div class="gobierto-data-sql-editor-modified-label-container">
+        <span class="gobierto-data-sql-editor-modified-label">
+          Consulta guardada
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue]: I18n or die

File: app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue [hunk 19]
```diff
@@ -97,50 +101,108 @@ export default {
...
+        this.labelValue = newValue
+      }
+    },
+    $route(to, from) {
+      if (to.path !== from.path) {
+        this.isSavingPromptVisible = false
+      }
+    },
+    showPrivate(newValue) {
+      this.isPrivate = newValue ? true : false
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue]:
```suggestion
      this.isPrivate = (newValue);
```

File: app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue [hunk 20]
```diff
@@ -97,50 +101,108 @@ export default {
...
     },
-    onClickEditHandler() {
-      this.isSavingPromptVisible = true
-      this.$nextTick(() => this.$refs.inputText.focus());
+    saveHandlerNewQuery() {
+      if (!this.isSavingPromptVisible) {
+        this.isSavingPromptVisible = true
+        this.$nextTick(() => this.$refs.inputText.focus());
+      } else {
+        if (this.labelValue === null) {
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue]:
```suggestion
        if (!this.labelValue) {
```

File: app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue [hunk 21]
```diff
@@ -97,50 +101,108 @@ export default {
...
     onKeyDownTextHandler(event) {
       const { value } = event.target
       this.labelValue = value
+      this.$root.$emit('enableSavedButton')
     },
     onInputCheckboxHandler(event) {
       const { checked } = event.target
       this.isPrivate = checked
     },
+    revertQuery() {
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue]: Add Handler names to user events

File: app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue [hunk 22]
```diff
@@ -97,50 +101,108 @@ export default {
...
+      this.$root.$emit('revertSavedQuery', true)
+      this.isSavingPromptVisible = false
+    },
+    disableSavingPrompt() {
+      this.isSavingPromptVisible = false
+      this.labelValue = null
+    }
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/SavingDialog.vue]: Remove this

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorCode.vue [hunk 23]
```diff
@@ -72,6 +88,18 @@ export default {
...
+      if (newValue !== oldValue) {
+        this.setEditorValue(this.queryDefault);
+        this.editor.focus()
+      }
+    },
+    revertQuerySaved(newValue, oldValue) {
+      if (newValue !== oldValue) {
+        this.setEditorValue(this.queryRevert);
+        this.editor.focus()
+      }
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorCode.vue]: Try to simplify this logic, printing only the queryStored for this component. This changes are populated upstream.

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorCode.vue [hunk 24]
```diff
@@ -122,7 +150,11 @@ export default {
 
         // update the query while typing
         this.$root.$emit("setCurrentQuery", value);
-      }, 250);
+        this.$root.$emit('enableSavedButton')
+        this.$root.$emit("resetQuery", false);
+        this.$root.$emit("revertSavedQuery", false);
+        this.$root.$emit("disabledStringSavedQuery", false);
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorCode.vue]: Move this logic, perhaps the events become useless, to the Dataset component, since it's it who handles the logic

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue [hunk 25]
```diff
@@ -122,31 +128,48 @@ export default {
...
-      isRecentModalActive: false,
+      isRecentModalActive: false
     };
   },
   created() {
     // it has to be the same event (keydown) as SQLEditorCode
     document.addEventListener("keydown", this.keyboardShortcutsListener);
   },
-  beforeDestroy() {
+  deactivated() {
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue:177]: Rollback to destroy

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue [hunk 26]
```diff
@@ -197,6 +220,14 @@ export default {
     closeQueriesModal() {
       this.isQueriesModalActive = false;
     },
+    resetQuery() {
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue]: Handler me

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue [hunk 27]
```diff
@@ -197,6 +220,14 @@ export default {
...
+      this.$refs.savingDialog.disableSavingPrompt();
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue]: Update the SavingDialog state via props, since such function is only triggered here.

**Review by Crashillo — CHANGES_REQUESTED** [review]

File: app/javascript/gobierto_data/webapp/components/Dataset.vue [hunk 28]
```diff
@@ -446,6 +529,60 @@ export default {
...
     },
+    getColumnsQuery(csv = '') {
+      const [ columns = '' ] = csv.split("\n");
+      this.arrayColumnsQuery = columns.split(",");
+    },
+    resetQuery(value) {
+      this.resetQueryDefault = value
+      if (value === true) {
+        this.isSavingPromptVisible = false
+        this.currentQuery = `SELECT * FROM ${this.tableName} LIMIT 50`;
```
Comment (ferblape) [inline app/javascript/gobierto_data/webapp/components/Dataset.vue:547]: The default query is duplicated in 3 different places. Can we move it to a single place? (method?)

File: app/javascript/gobierto_data/webapp/components/Dataset.vue [hunk 29]
```diff
@@ -399,8 +471,19 @@ export default {
         this.items = items;
         this.queryDuration = new Date().getTime() - startTime;
         this.isQueryRunning = false;
+        this.getColumnsQuery(this.items)
```
Comment (ferblape) [inline app/javascript/gobierto_data/webapp/components/Dataset.vue:480]: You should add `this.queryError = null` to clear errors when the query runs succesfully

**Review by ferblape — CHANGES_REQUESTED** [review]
Just a couple of details and for me is ready to release

File: app/javascript/gobierto_data/webapp/components/Dataset.vue (hunk 28 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_data/webapp/components/Dataset.vue:547]: I remove one of them. I can't remove `currenTquery` in `resetQuery` method because `currentQuery` is continuously updated and listen to any changes in the editor, so when reset the editor to initial state we need resetQuery to `SELECT * FROM ${this.tableName} LIMIT 50`

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue (hunk 25 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue:177]: if we use `beforeDestroy` instead of  `deactivated` this bug appears again: https://github.com/PopulateTools/issues/issues/1013

Comment (Crashillo) [reply app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue:177]: If #1013 happens, that means `SQLEditorHeader` is not being destroyed. And that happens because `SQLEditorHeader` is wrapped in a `<keep-alive>` clause. Is that right?
Then, we've to look why is SQLEditorHeader kept alive. It shouldn't. It only belongs to DataTab component, who is in charge to create it. Is DataTab kept alive also? Why reason?

Comment (Crashillo) [reply app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorHeader.vue:177]: The point of this is about the lifecycles. `created` <-> `destroyed`. `activated` <-> `deactivated`. But never mixed.

**Review by ferblape — APPROVED** [review]

File: app/javascript/gobierto_data/webapp/components/Dataset.vue [hunk 30]
```diff
@@ -149,20 +165,36 @@ export default {
...
+  beforeRouteEnter (to, from, next) {
+    const {
+      name: nameComponent
+    } = to;
+    next(vm => {
+      if (nameComponent === 'Query') {
+        vm.showRevertQuery = true
+      } else {
+        vm.showRevertQuery = false
+      }
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/Dataset.vue]:
```suggestion
   vm.showRevertQuery = (nameComponent === 'Query')
```

File: app/javascript/gobierto_data/webapp/components/Dataset.vue [hunk 31]
```diff
@@ -263,16 +316,46 @@ export default {
...
+      //Find the query
+      const getSavedQuery = items.find(({ id }) => id === queryId) || {}
+
+      //We need to keep this query separate from the editor query
+      //When load a saved query we use the queryId to find inside privateQueries or publicQueries
+      const {
+        attributes: {
+          sql: queryRevert
+        } = {}
+      } = getSavedQuery
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/Dataset.vue]: Simplify

File: app/javascript/gobierto_data/webapp/components/Dataset.vue (hunk 28 above)
Comment (Crashillo) [reply app/javascript/gobierto_data/webapp/components/Dataset.vue:547]: Explore the idea about setting a `computed` prop, something like, `this.queryDatasetDefault`

**Review by Crashillo — CHANGES_REQUESTED** [review]

File: app/javascript/gobierto_data/webapp/components/Dataset.vue (hunk 28 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_data/webapp/components/Dataset.vue:547]: I tried, but it didn't work

**Review by Crashillo — APPROVED** [review]
